src/cpp/src/sampler.cpp

```diff
@@ -587,45 +662,67 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
+                        
+                        // do sampling only for token validation/generation.
+                        // continue in case of extending draft model sequences by main model generated tokens which
+                        // should be taken to KV cache without validation
+                        if (!is_validation_mode_enabled && token_offset > 0) {
+                            continue;
+                        }
+
+                        auto logit_vector = _get_logit_vector(sequence_group_logits, running_sequence_id, token_offset);
+                        logit_processor.apply(logit_vector);
+
+                        Token sampled_token_id;
+                        if (sampling_params.is_greedy_decoding()) {
+                            sampled_token_id = _greedy_sample(logit_vector);
+                        } else {
+                            // is_multinomial()
+                            const bool is_generate_n_tokens = sequence_group->num_total_seqs() == 1;
+                            const size_t num_tokens_per_sequence = is_generate_n_tokens ? sampling_params.num_return_sequences : 1;
+                            auto sampled_token_ids = _multinomial_sample(logit_vector, num_tokens_per_sequence);
+                            OPENVINO_ASSERT(sampled_token_ids.size(), num_tokens_per_sequence);
+                            if (is_generate_n_tokens) {
+                                const auto forked_seq_ids = create_n_forked_sequences(sequence_group, logit_processor, sampled_token_ids);
+                                sampler_output.m_forked_sequences.insert({running_sequences[0]->get_id(), forked_seq_ids});
                             }
-                            sampler_output.m_forked_sequences.insert({running_sequences[0]->get_id(), forked_seq_ids});
+                            sampled_token_id = sampled_token_ids.front();
+                        }
+                        // flag to add sampled token to generated sequence or extend logit processors only
+                        bool is_extend_sequence = token_offset == 0,
+                             // flag to update generated length of sequence group in logit processor
+                             is_update_len_logit_processor = running_sequence_id == num_running_sequences - 1,
+                             is_validation_failed = true;
+                        if (is_validation_mode_enabled) {
+                            is_validation_failed = validate_candidate(running_sequences[running_sequence_id], token_offset, sampled_token_id, is_extend_sequence, decrease_context_len_per_seq_group);
+                        }
+                        register_new_token(sampled_token_id, running_sequences[running_sequence_id], logit_processor, is_extend_sequence, is_update_len_logit_processor);
+                        // to exit from sampling in case of failed token validation
+                        if (!is_validation_failed) {
```

Comment:
```suggestion
                        if (is_validation_failed) {
```
should it be corrected like this?

Reply:
fixed